### DISCUSSION TITLE
Weird behavior of multiple shadow-casting spot lights

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3872] [Q&A] [answered]
I have a test program with 3 shadow-casting spot lights, identically configured except for intensity.  If I have spotlight 1 set to high intensity (1000000 lumens) and spotlights 2 and 3 turned way down to 0.1 lumens, I get a good shadow.  Same if spotlight 2 is bright and 1 and 3 are dim.  But if spotlight 3 is bright and 1 and 2 are dim, it casts no shadow.  At that point, one might think there is something wrong with spotlight 3.  But if I then remove spotlight 1 and/or spotlight 2 from the scene, there's a shadow.  Does this make any sense to anyone?  If not, I'll try to construct a proper bug report.

**Comment by romainguy** [comment, marked as answer]
There is a limit to the number of spot lights that can cast shadows at the same time in the frustum. @bejado can give you the details. It might just be that we cast shadows for the first 2 spotlights in the scene so 3, even when bright, doesn't get one. We've been thinking of giving priority to lights closer to the camera or the brighter lights (it should really be a combo of both).

**Reply by jwwalker** [reply]
Ouch.  If there can be only 2 shadow-casting spot lights at the same time in the frustum, that would kill my evil plot to simulate a shadow-casting point light using 4 shadow-casting spot lights.

**Comment by jwwalker** [comment]
@bejado Is that correct that you can't have more than 2 shadow-casting spot lights visible at a time?  I was hoping to use Filament to replace OpenGL in my product, but this shadow limitation might be a deal-breaker.

**Reply by bejado** [reply]
Yes, to be compatible on a wide array of platforms, we currently have the limit set to 2 shadow-casting spot lights: https://github.com/google/filament/blob/main/libs/filabridge/include/private/filament/EngineEnums.h#L52-L56

The GPU you're targeting may be able to support more. You'd just need to update that constant and build Filament from source.

**Reply by romainguy** [reply]
@bejado If we moved the computation to the fragment shader we wouldn't be limited anymore right? I wonder if we should just bite that bullet (or maybe bite it only for shadows > 2).

**Reply by jwwalker** [reply]
@bejado Apparently I can't just boost up `CONFIG_MAX_SHADOW_CASTING_SPOTS` as much as I want.  In filament/src/details/View.h, there is a comment "Because we're using a uint8_t for the visibility mask, we're limited to 6 spot light shadows" and a static assert enforcing that limit.

**Reply by jwwalker** [reply]
I tried increasing `CONFIG_MAX_SHADOW_CASTING_SPOTS` to 6 and recompiling Filament, then running a test with 3 spot lights.  I saw no shadows, but did see weird black specks dancing around the image.

**Reply by bejado** [reply]
What backend/gpu are you targeting?
@romainguy Yeah, we could move the computation for light > 2 to the fragment shader. We'll still be limited to the 6 spot light shadows for now.

**Reply by jwwalker** [reply]
@bejado I'm using Metal and Radeon Pro 560X 4 GB.

Any chance of getting more than 6 spot light shadows?  What's the reason for that limit?

**Reply by bejado** [reply]
I'm surprised Metal isn't allowing more variants; I'll look into that.

As the comment alludes to, we use a `uint8_t` to track visibility. 2 of these bits are used for other purposes, which leaves 6 to track spot light shadows. We could potentially update this to a wide format, or make it configurable.

**Reply by jwwalker** [reply]
@bejado I tried changing my test program to use an OpenGL back end, and with `CONFIG_MAX_SHADOW_CASTING_SPOTS = 6`, I still can't get more than 2 shadow-casting spot lights to work.

**Reply by jwwalker** [reply]
With the OpenGL backend and `CONFIG_MAX_SHADOW_CASTING_SPOTS = 6`, three spot lights don't give me the black specks I saw with Metal, but slightly blotchy appearance as if the object had gotten wet unevenly:

![three spots, shadows](https://user-images.githubusercontent.com/730426/116946123-3144d600-ac2e-11eb-8e1f-3cc953cdaaf6.png)

For comparison, here's the same scene with 2 spot lights:

![two spots, shadows](https://user-images.githubusercontent.com/730426/116946152-46ba0000-ac2e-11eb-94b4-05b5ff43fa20.png)

and here's 3 spot lights but none of them set to cast shadows:

![three spots, no shadow](https://user-images.githubusercontent.com/730426/116946184-56394900-ac2e-11eb-9ba4-6c47aee6c3e5.png)

**Reply by bejado** [reply]
This is a bit strange, I'm unable to reproduce on my Radeon Pro 5500M. With both the OpenGL and Metal backends I'm successfully able to render 6 spot light shadows without artifacts. The Radeon Pro 560X is a few years older, yet according to the [Metal Feature Set Tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf), all Mac GPUs should support the same number of varying components (124).

I suspect you're somehow hitting that limit, though even with 6 spot light shadows we're nowhere close. Can you provide the Filament material you're using to render?

**Reply by bejado** [reply]
You can also try building the [bjd/spot-shadow-varyings](https://github.com/google/filament/tree/bjd/spot-shadow-varyings) branch, which removes all varyings related to spot shadows. If it still doesn't work for you, then something else must be going on.

**Reply by jwwalker** [reply]
@bejado Thanks for checking.  I built the spot-shadow-varyings branch and still had problems.  I guess it's time for me to start trying to build a minimal example.

**Reply by bejado** [reply]
You could also try to reproduce in our `gltf_viewer` sample, or use it as a starting point.

**Reply by jwwalker** [reply]
@bejado OK, I figured out what I was doing wrong.  Although I was linking with libraries built with `CONFIG_MAX_SHADOW_CASTING_SPOTS = 6`, I was baking the material using an old build of `matc`.

If the number of shadows could be made more configurable, that would be great.  Even 6 isn't really enough for my purposes.  I suppose I could consider a multi-pass approach where I render to textures with different batches of lights, and then add the images together, but that would be ugly.

**Reply by bejado** [reply]
Yes, one of our goals is to make the number of allowed spot shadows configurable.
